core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java

```diff
@@ -1082,6 +1085,7 @@ public Optional<Type> getSupportedType(Session session, CatalogHandle catalogHan
         CatalogMetadata catalogMetadata = getCatalogMetadata(session, catalogHandle);
         ConnectorMetadata metadata = catalogMetadata.getMetadata(session);
         return metadata.getSupportedType(session.toConnectorSession(catalogHandle), tableProperties, type)
+                .or(() -> this.getSupportedTypeForStructuralTypes(session, catalogHandle, tableProperties, type))
```

Comment:
I am thinking that your solution technically works.
The approach you are taking I'm not 100% ok with.
We're asking the connector to give us the supported type and if the call returns `Optional.empty()` we have a custom handling which asks (depending on the structural type yet again the connector to give the support type).

I understand the reasoning to avoid code redundancy.

I'd argue that adding a bunch of default methods in the `ConnectorMetadata` interface to handle the structural types would be solving the problem.

Reply:
Is it possible to move this logic to the Connector ? What if the connector supports coercion for a specific type but not for structural type ?

Reply:
The same logic can be moved to Iceberg or DeltaLake connector which performs a simple check today.